### DISCUSSION TITLE
Paparazzi 2.0 Design

### DISCUSSION THREAD

**Discussion by geoff-powell** [cashapp/paparazzi#1711] [Ideas]
👋🏽 Hello, 

This discussion will be used to ideate over the next API design of paparazzi. 

# WIP

Old 2.0 Refactor
[Old snapshotter branch](https://github.com/BrianGardnerAtl/paparazziSplit/commit/95e0a2b5aaafc0042433eec25900fac8a30ccd32#diff-f06b2b0cccf885fca45e37f4e32903aa7e6958c0934616739a0a7c18d7f33149R156)

Brainstorming a way to separate our snapshotter interface from the component that renders the views. This could provide a way to allow Paparazzi snapshotting in Robolectric or Espresso tests (KMP when?).
```
(interceptor pattern)

interface Snapshotter {
  fun snapshot(): Snapshot
  fun clip(): Clip // Video
}


// Could be future iteration?
(multi-platform apis for other snapshot producers)

// Eventually have LayoutLibEngine, RobolectricEngine, EspressoEngine, SwiftUiEngine, etc.
interface Engine {
  fun render(View): BufferedImage
}



// test module
interface TestSnapshotHandler {
  fun handleSnapshot(Snapshot, TestRecord)
  fun handleClip(Clip, TestRecord)
}

class Paparazzi(...) : TestRule {}}
```

### Keyframe API example
Adds ability to take actions at different points of the video being captured
```
paparazzi.gif(
      view = container,
      start = 0L,
      end = 2_000L,
      keyframes = listOf(
        // t=200: start loading. Transition will finish at t=500. (start + 300).
        Keyframe(200L) { loadingHelper.isLoading = true },

        // t=400: request stop loading. It will schedule this for t=1200 (start + 1000).
        Keyframe(400L) { loadingHelper.isLoading = false },

        // t=1600: start loading again. Will finish at t=1900.
        Keyframe(1_600L) { loadingHelper.isLoading = true },
      ),
    )
```
### TODOs

* Look into SSIM, FLIP diffing algorithms
* Paparazzi annotation accessibility support
* Paparazzi code-gen manifest useful for Google Compose snapshots

* SDK-ify work needs to be done generating R.jar 
* SDK paparazzi values need to come from somewhere
* Jessie's Keyframe API (above)

# Meeting Notes
### Agenda [1/22/25]
* 2.0 Design thoughts: 
* Want to cut a 2.0.0 release in next 2 weeks
* Merge in annotation processing
* Update on sdkification of paparazzi
* Font resource loading: https://github.com/cashapp/paparazzi/pull/1798 : solved via [comment](https://github.com/DSteve595)

### Agenda [5/30/25]
* Paparazzi out of Junit - expect help to reach finish line
* 2.0.0 alpha02 release to unblock community
* Compose 1.8.0 regression fixes - ripple choreographer timings, accessibility snapshot bounding box rendering index is incorrect
* FakeSystemUI work from @saket 
** Layouts breaking with insets in CashApp
** Add Touch inputs to record
** Test container animation with paparazzi and touch integration

* Create paparazzi snapshot browser for free similar to Square Spoon, Cash Snapshot Browser

**Comment by Ashl7** [comment]
I think @saket brought up this idea, a `.waitForIdle()` api similar to android tests would be nice, since paparazzi usually captures screenshot on first frame. Happy to contribute to this.

**Reply by geoff-powell** [reply]
Ya unfortunately this is a tough api to support. I know there is the `InstantAnimationRule` for legacy views but I would argue the correct approach is:

* Use `pararazzi.gif()` and record video of expected output
* Use `paparazzi.snapshot(offsetMillis = {offsetToDesiredState})` to take snapshot of your expected state. (I know there is a caveat here that this doesn't work for compose animations at the moment)

**Reply by saket** [reply]
While I agree that supporting this will require a larger refactor, I don't think using offsets is the correct approach, as it relies on knowing the correct offset. I would never be able to calculate it accurately and would have to resort to guessing random numbers until my screenshots look correct.

**Comment by TWiStErRob** [comment]
Rename historical `paparazzi.gif` to express what it does instead of expressing what file format it [does not](https://github.com/cashapp/paparazzi/pull/1186) save. Some ideas:
* `.snapshotAnimation()`
* `.animatedSnapshot()`
* `.movingSnapshot()` (feels a bit contradictory, as "snapshot" is a still photograph as far as I understand)
* `.video()`
* `.animation()`
* `.snapshot()` (i.e. just rely on overload with named params)
